hazelcast/src/main/java/com/hazelcast/query/impl/AttributeIndexRegistry.java

```diff
@@ -0,0 +1,320 @@
+ * composite indexes to allow transparent querying on their first components.
+ */
+public class AttributeIndexRegistry {
+
+    private final Map<String, Record> registry = new ConcurrentHashMap<String, Record>();
+
+    /**
+     * Registers the given index in this registry.
+     * <p>
+     * This method expects external thread access synchronization applied, so
+     * there is no more than one writer at any given time.
+     *
+     * @param index the index to register.
+     * @see Indexes#addOrGetIndex
+     */
+    public void register(InternalIndex index) {
+        String[] components = index.getComponents();
+        String attribute = components == null ? index.getName() : components[0];
+
+        Record record = registry.get(attribute);
+        if (record == null) {
+            record = new Record();
+            registry.put(attribute, record);
+        }
+
+        if (index.isOrdered()) {
+            if (record.orderedWorseThan(index)) {
+                record.ordered = components == null ? index : new FirstComponentDecorator(index);
+            }
+        } else {
+            if (record.unorderedWorseThan(index)) {
+                record.unordered = index;
+            }
+        }
+    }
+
+    /**
+     * Matches an index for the given attribute and match hint.
+     *
+     * @param attribute the attribute to match an index for.
+     * @param matchHint the match hint; {@link QueryContext.IndexMatchHint#EXACT_NAME}
+     *                  is not supported by this method.
+     * @return the matched index or {@code null} if nothing matched.
+     * @see QueryContext.IndexMatchHint
+     */
+    public InternalIndex match(String attribute, QueryContext.IndexMatchHint matchHint) {
+        Record record = registry.get(attribute);
+        if (record == null) {
+            return null;
+        }
+
+        switch (matchHint) {
+            case NONE:
+                // Intentional fallthrough. We still prefer ordered indexes
+                // under the cover since they are more universal in terms of
+                // supported fast queries.
+            case PREFER_ORDERED:
+                InternalIndex ordered = record.ordered;
+                return ordered == null ? record.unordered : ordered;
+            case PREFER_UNORDERED:
+                InternalIndex unordered = record.unordered;
+                return unordered == null ? record.ordered : unordered;
+            default:
+                throw new IllegalStateException("unexpected match hint: " + matchHint);
+        }
+    }
+
+    /**
+     * Clears this registry by unregistering all previously registered indexes.
+     */
+    public void clear() {
+        registry.clear();
+    }
+
+    private static class Record {
+
+        // It's enough to have volatile non-atomic mutations here, just to ensure
+        // the visibility, since only a single thread may mutate a record at any
```

Comment:
can we have the answer of `which was the mutator single thread` in this comment?